File: mobile/src/library/sliceOverrides.ts
// Pure builders for per-slice parameter overrides. Bambuddy's slice API takes only preset
// REFERENCES — but local presets support `inherits` + a delta of keys (Bambuddy resolves the base;
// the slicer sidecar flattens inheritance too). So "advanced mode" = upsert an ephemeral local
// preset carrying just the user's changed keys, then slice referencing it. Same mechanism the
// "+ Supports" twins use (deploy/bambuddy/ensure-support-profiles.py), generalized.
//
// Key names + value shapes verified against BambuStudio's PrintConfig.cpp: preset JSON values are
// STRINGS ("1"/"0" bools, "15%" percents). Only scalar-safe keys are exposed — per-extruder ARRAY
// keys (speeds: 5-element (extruder,variant) arrays on H2-series) are deliberately excluded until
// the app can read base preset content to rewrite whole arrays.

export interface SliceOverrides {
  /** wall_loops (0–1000; stock ~2) */
  wallLoops?: number;
  /** sparse_infill_density percent 0–100 */
  infillDensity?: number;
  /** sparse_infill_pattern */
  infillPattern?: string;
  /** top_surface_pattern */
  topPattern?: string;
  /** enable_prime_tower */
  primeTower?: boolean;
  /** prime_tower_width mm (min 2) */
  primeTowerWidth?: number;
  /** enable_support — advanced path (supersedes the "+ Supports" twin swap when set) */
  support?: boolean;
  /** support_type: normal(auto) | tree(auto) | normal(manual) | tree(manual) */
  supportType?: string;
  /** support_style: default | grid | snug | tree_slim | tree_strong | tree_hybrid | tree_organic */
  supportStyle?: string;
  /** support_threshold_angle 1–90° */
  supportAngle?: number;
  /** filament_flow_ratio (FILAMENT preset; H2-series carries a 3-element per-variant array) */
  flowRatio?: number;
}

export const INFILL_PATTERNS = ['grid', 'gyroid', 'cubic', 'triangles', 'honeycomb', 'lightning', 'adaptivecubic', 'crosshatch'] as const;
export const TOP_PATTERNS = ['monotonic', 'monotonicline', 'concentric', 'alignedrectilinear'] as const;
export const SUPPORT_TYPES = ['tree(auto)', 'normal(auto)'] as const;
export const SUPPORT_STYLES = ['default', 'snug', 'tree_slim', 'tree_strong', 'tree_hybrid', 'tree_organic'] as const;

const PROCESS_KEYS: (keyof SliceOverrides)[] = ['wallLoops', 'infillDensity', 'infillPattern', 'topPattern', 'primeTower', 'primeTowerWidth', 'support', 'supportType', 'supportStyle', 'supportAngle'];

export function hasProcessOverrides(o: SliceOverrides): boolean {
  return PROCESS_KEYS.some((k) => o[k] !== undefined);
}
export function hasFilamentOverrides(o: SliceOverrides): boolean {
  return o.flowRatio !== undefined;
}

const bool = (b: boolean): string => (b ? '1' : '0');

/**
 * Delta process-preset `setting` JSON inheriting `baseQualityName`, or null when nothing to
 * override. `presetName` is the reusable local-preset row name (one per machine, updated in place).
 */
export function buildProcessDelta(baseQualityName: string, o: SliceOverrides, presetName: string): Record<string, string> | null {
  if (!hasProcessOverrides(o)) return null;
  const s: Record<string, string> = { type: 'process', name: presetName, from: 'User', inherits: baseQualityName };
  if (o.wallLoops !== undefined) s.wall_loops = String(Math.max(0, Math.round(o.wallLoops)));
  if (o.infillDensity !== undefined) s.sparse_infill_density = `${Math.min(100, Math.max(0, Math.round(o.infillDensity)))}%`;
  if (o.infillPattern !== undefined) s.sparse_infill_pattern = o.infillPattern;
  if (o.topPattern !== undefined) s.top_surface_pattern = o.topPattern;
  if (o.primeTower !== undefined) s.enable_prime_tower = bool(o.primeTower);
  if (o.primeTowerWidth !== undefined) s.prime_tower_width = String(Math.max(2, o.primeTowerWidth));
  if (o.support !== undefined) s.enable_support = bool(o.support);
  if (o.supportType !== undefined) s.support_type = o.supportType;
  if (o.supportStyle !== undefined) s.support_style = o.supportStyle;
  if (o.supportAngle !== undefined) s.support_threshold_angle = String(Math.min(90, Math.max(1, Math.round(o.supportAngle))));
  return s;
}

/**
 * Delta filament-preset `setting`, or null. `variants` = the machine's per-(extruder,variant) array
 * length for filament keys (H2-series: 3; single-extruder: 1) — same value replicated across, which
 * is the safe uniform override.
 */
export function buildFilamentDelta(baseFilamentName: string, o: SliceOverrides, presetName: string, variants = 3): Record<string, unknown> | null {
  if (!hasFilamentOverrides(o)) return null;
  const flow = Math.min(2, Math.max(0.5, o.flowRatio!));
  return {
    type: 'filament',
    name: presetName,
    from: 'User',
    inherits: baseFilamentName,
    filament_flow_ratio: Array.from({ length: Math.max(1, variants) }, () => String(flow)),
  };
}

/** Count of active overrides — drives the "n changed" badge on the Advanced accordion. */
export function overrideCount(o: SliceOverrides): number {
  return [...PROCESS_KEYS, 'flowRatio' as const].filter((k) => o[k] !== undefined).length;
}

File: native/Sprout/Domain/SliceOverrides.swift
import Foundation

// Pure builders for per-slice parameter overrides. Bambuddy's slice API takes only preset
// REFERENCES — but local presets support `inherits` plus a delta of keys (Bambuddy resolves the
// base; the slicer sidecar flattens inheritance too). So "advanced mode" = upsert an ephemeral
// local preset carrying just the user's changed keys, then slice referencing it. Same mechanism
// the "+ Supports" twins use (deploy/bambuddy/ensure-support-profiles.py), generalized.
//
// Key names and value shapes are verified against BambuStudio's PrintConfig.cpp: preset JSON
// values are STRINGS ("1"/"0" bools, "15%" percents). Only scalar-safe keys are exposed —
// per-extruder ARRAY keys (speeds are 5-element (extruder,variant) arrays on the H2-series) are
// deliberately excluded until the app can read base preset content and rewrite whole arrays.

/// The user's per-slice parameter changes. Every field is optional: nil means "inherit the base
/// preset", which is what keeps the emitted delta down to only what actually changed.
struct SliceOverrides: Hashable, Sendable {
    /// `wall_loops` (0–1000; stock is ~2).
    var wallLoops: Double?
    /// `sparse_infill_density`, a percentage 0–100.
    var infillDensity: Double?
    /// `sparse_infill_pattern`.
    var infillPattern: String?
    /// `top_surface_pattern`.
    var topPattern: String?
    /// `enable_prime_tower`.
    var primeTower: Bool?
    /// `prime_tower_width` in mm (the slicer's minimum is 2).
    var primeTowerWidth: Double?
    /// `enable_support` — the advanced path, which supersedes the "+ Supports" twin swap when set.
    var support: Bool?
    /// `support_type`: normal(auto) | tree(auto) | normal(manual) | tree(manual).
    var supportType: String?
    /// `support_style`: default | grid | snug | tree_slim | tree_strong | tree_hybrid | tree_organic.
    var supportStyle: String?
    /// `support_threshold_angle`, 1–90°.
    var supportAngle: Double?
    /// `filament_flow_ratio`. This one lives on the FILAMENT preset, not the process preset, and
    /// the H2-series stores it as a 3-element per-variant array.
    var flowRatio: Double?
}

// MARK: - Pickable values

extension SliceOverrides {
    static let infillPatterns: [String] = [
        "grid", "gyroid", "cubic", "triangles", "honeycomb", "lightning", "adaptivecubic", "crosshatch",
    ]
    static let topPatterns: [String] = ["monotonic", "monotonicline", "concentric", "alignedrectilinear"]
    static let supportTypes: [String] = ["tree(auto)", "normal(auto)"]
    static let supportStyles: [String] = [
        "default", "snug", "tree_slim", "tree_strong", "tree_hybrid", "tree_organic",
    ]
}

// MARK: - Which keys are set

extension SliceOverrides {
    /// One flag per PROCESS key, listed once so `hasProcessOverrides` and `overrideCount` can never
    /// disagree about what counts as a process override. `flowRatio` is absent on purpose: it goes
    /// on the filament preset, not this one.
    private var processKeysPresent: [Bool] {
        [
            wallLoops != nil, infillDensity != nil, infillPattern != nil, topPattern != nil,
            primeTower != nil, primeTowerWidth != nil, support != nil, supportType != nil,
            supportStyle != nil, supportAngle != nil,
        ]
    }

    var hasProcessOverrides: Bool { processKeysPresent.contains(true) }

    var hasFilamentOverrides: Bool { flowRatio != nil }

    /// Count of active overrides — drives the "n changed" badge on the Advanced accordion.
    var overrideCount: Int {
        processKeysPresent.filter { $0 }.count + (flowRatio != nil ? 1 : 0)
    }
}

// MARK: - Delta builders

extension SliceOverrides {
    /// The delta process-preset `setting` payload inheriting `baseQualityName`, or nil when nothing
    /// is overridden (so an untouched slice causes no preset churn at all).
    ///
    /// `presetName` is the reusable local-preset row name — one row per machine, updated in place,
    /// so slicing does not accumulate a preset row per job.
    func processDelta(inheriting baseQualityName: String, presetName: String) -> [String: JSONValue]? {
        guard hasProcessOverrides else { return nil }
        var s: [String: JSONValue] = [
            "type": .string("process"),
            "name": .string(presetName),
            "from": .string("User"),
            "inherits": .string(baseQualityName),
        ]
        if let v = wallLoops {
            s["wall_loops"] = .string(String(clampedInt(v, lower: 0, upper: Int.max)))
        }
        if let v = infillDensity {
            s["sparse_infill_density"] = .string("\(clampedInt(v, lower: 0, upper: 100))%")
        }
        if let v = infillPattern { s["sparse_infill_pattern"] = .string(v) }
        if let v = topPattern { s["top_surface_pattern"] = .string(v) }
        if let v = primeTower { s["enable_prime_tower"] = .string(presetBool(v)) }
        if let v = primeTowerWidth {
            // No rounding here — a tower width is a real millimetre value, only floored at the
            // slicer's minimum.
            s["prime_tower_width"] = .string(presetNumber(max(2, v)))
        }
        if let v = support { s["enable_support"] = .string(presetBool(v)) }
        if let v = supportType { s["support_type"] = .string(v) }
        if let v = supportStyle { s["support_style"] = .string(v) }
        if let v = supportAngle {
            s["support_threshold_angle"] = .string(String(clampedInt(v, lower: 1, upper: 90)))
        }
        return s
    }

    /// The delta filament-preset `setting`, or nil when the flow ratio is untouched.
    ///
    /// `variants` is the machine's per-(extruder,variant) array length for filament keys —
    /// H2-series: 3, single-extruder: 1. The same value is replicated across every slot, which is
    /// the safe uniform override.
    func filamentDelta(
        inheriting baseFilamentName: String,
        presetName: String,
        variants: Int = 3
    ) -> [String: JSONValue]? {
        guard let ratio = flowRatio else { return nil }
        let flow = presetNumber(min(2, max(0.5, ratio)))
        return [
            "type": .string("filament"),
            "name": .string(presetName),
            "from": .string("User"),
            "inherits": .string(baseFilamentName),
            "filament_flow_ratio": .array(Array(repeating: .string(flow), count: max(1, variants))),
        ]
    }
}

// MARK: - Value formatting

/// BambuStudio's `"1"` / `"0"` boolean convention for preset JSON.
private func presetBool(_ b: Bool) -> String { b ? "1" : "0" }

/// Renders a number the way preset JSON expects it: an integral value must serialise as "2", never
/// "2.0", because the slicer compares these values as text. Non-integral values keep the shortest
/// form that round-trips. Interpolation is used rather than a formatter so the output can never
/// pick up a comma decimal separator from the device locale.
private func presetNumber(_ v: Double) -> String {
    // The magnitude bound also rejects NaN and infinity, either of which would trap `Int64(_:)`.
    if v == v.rounded(), abs(v) < 1e15 { return String(Int64(v)) }
    return "\(v)"
}

/// Round to the nearest integer and clamp into `lower...upper` without trapping.
///
/// The clamp is done in Double space on purpose: `Int(_:)` on a Double outside Int's range is a
/// hard crash, and a numeric text field bound to one of these sliders can produce exactly that.
private func clampedInt(_ v: Double, lower: Int, upper: Int) -> Int {
    guard v.isFinite else { return lower }
    let r = v.rounded()
    if r <= Double(lower) { return lower }
    if r >= Double(upper) { return upper }
    return Int(r)
}
